File: app/src/main/java/org/mozilla/focus/search/ManualAddSearchEnginePreference.java
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.search;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Parcel;
import android.os.Parcelable;
import android.preference.Preference;
import android.support.design.widget.TextInputLayout;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ProgressBar;

import org.mozilla.focus.R;
import org.mozilla.focus.utils.UrlUtils;

public class ManualAddSearchEnginePreference extends Preference {
    private EditText engineNameEditText;
    private EditText searchQueryEditText;
    private TextInputLayout engineNameErrorLayout;
    private TextInputLayout searchQueryErrorLayout;
    private ProgressBar progressView;

    private String querySearchString;
    private String engineNameString;

    public ManualAddSearchEnginePreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ManualAddSearchEnginePreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        final Parcelable superState = super.onSaveInstanceState();

        final PreferenceSavedState preferenceSavedState = new PreferenceSavedState(superState);
        preferenceSavedState.engineName = engineNameEditText.getText().toString();
        preferenceSavedState.searchQuery = searchQueryEditText.getText().toString();

        return preferenceSavedState;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        PreferenceSavedState preferenceSavedState = (PreferenceSavedState) state;
        super.onRestoreInstanceState(preferenceSavedState.getSuperState());

        engineNameString = preferenceSavedState.engineName;
        querySearchString = preferenceSavedState.searchQuery;
    }

    @Override
    protected View onCreateView(ViewGroup parent) {
        final View view = super.onCreateView(parent);

        engineNameErrorLayout = view.findViewById(R.id.edit_engine_name_layout);
        searchQueryErrorLayout = view.findViewById(R.id.edit_search_string_layout);

        engineNameEditText = view.findViewById(R.id.edit_engine_name);
        searchQueryEditText = view.findViewById(R.id.edit_search_string);

        restoreViewState();

        progressView = view.findViewById(R.id.progress);

        engineNameEditText.addTextChangedListener(buildTextWatcherForErrorLayout(engineNameErrorLayout));
        searchQueryEditText.addTextChangedListener(buildTextWatcherForErrorLayout(searchQueryErrorLayout));

        return view;
    }

    /**
     * Restores the saved state of the UI
     */
    private void restoreViewState() {
        if (!TextUtils.isEmpty(engineNameString)) {
            engineNameEditText.setText(engineNameString);
        }
        if (!TextUtils.isEmpty(querySearchString)) {
            searchQueryEditText.setText(querySearchString);
        }
    }

    private boolean engineNameIsUnique(String engineName) {
        final SharedPreferences sharedPreferences = getContext().getSharedPreferences(SearchEngineManager.PREF_FILE_SEARCH_ENGINES, Context.MODE_PRIVATE);
        return !sharedPreferences.contains(engineName);
    }

    public boolean validateEngineNameAndShowError(String engineName) {
        if (TextUtils.isEmpty(engineName) || !engineNameIsUnique(engineName)) {
            engineNameErrorLayout.setError(getContext().getString(R.string.search_add_error_empty_name));
            return false;
        } else {
            engineNameErrorLayout.setError(null);
            return true;
        }
    }

    public boolean validateSearchQueryAndShowError(String searchQuery) {
        if (TextUtils.isEmpty(searchQuery)) {
            searchQueryErrorLayout.setError(getContext().getString(R.string.search_add_error_empty_search));
            return false;
        } else if (!UrlUtils.isValidSearchQueryUrl(searchQuery)) {
            searchQueryErrorLayout.setError(getContext().getString(R.string.search_add_error_format));
            return false;
        } else {
            searchQueryErrorLayout.setError(null);
            return true;
        }
    }

    private TextWatcher buildTextWatcherForErrorLayout(final TextInputLayout errorLayout) {
        return new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {}

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                errorLayout.setError(null);
            }

            @Override
            public void afterTextChanged(Editable editable) {}
        };
    }

    public void setSearchQueryErrorText(final String err) {
        searchQueryErrorLayout.setError(err);
    }

    public void setProgressViewShown(final boolean isShown) {
        progressView.setVisibility(isShown ? View.VISIBLE : View.GONE);
    }

    /**
     * Class for saving the state of the UI - search engine name and search string
     */
    private static class PreferenceSavedState extends Preference.BaseSavedState {
        private String engineName;
        private String searchQuery;

        PreferenceSavedState(Parcelable superState) {
            super(superState);
        }

        PreferenceSavedState(Parcel source) {
            super(source);
            // Get the current preference's value
            engineName = source.readString();
            searchQuery = source.readString();
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            super.writeToParcel(dest, flags);
            // Write the preference's value
            dest.writeString(engineName);
            dest.writeString(searchQuery);
        }

        // Standard creator object using an instance of this class
        public static final Parcelable.Creator<PreferenceSavedState>
                CREATOR =
                new Parcelable.Creator<PreferenceSavedState>() {

                    public PreferenceSavedState createFromParcel(
                            Parcel in) {
                        return new PreferenceSavedState(in);
                    }

                    public PreferenceSavedState[] newArray(int size) {
                        return new PreferenceSavedState[size];
                    }
                };
    }
}

File: app/src/main/java/org/mozilla/focus/search/ManualAddSearchEnginePreference.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.search

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.os.Parcelable
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.AttributeSet
import android.view.View
import android.widget.EditText
import android.widget.ProgressBar
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import com.google.android.material.textfield.TextInputLayout
import org.mozilla.focus.R
import org.mozilla.focus.utils.UrlUtils
import org.mozilla.focus.utils.ViewUtils
@Suppress("TooManyFunctions")
class ManualAddSearchEnginePreference(context: Context, attrs: AttributeSet) :
    Preference(context, attrs) {
    private var engineNameEditText: EditText? = null
    private var searchQueryEditText: EditText? = null
    private var engineNameErrorLayout: TextInputLayout? = null
    private var searchQueryErrorLayout: TextInputLayout? = null

    private var progressView: ProgressBar? = null

    private var savedSearchEngineName: String? = null
    private var savedSearchQuery: String? = null
    private var lastEditTextTouched: String? = null
    private var cursorPosition: Int? = null

    @SuppressLint("ClickableViewAccessibility")
    override fun onBindViewHolder(holder: PreferenceViewHolder?) {
        super.onBindViewHolder(holder)

        engineNameErrorLayout =
                holder!!.findViewById(R.id.edit_engine_name_layout) as TextInputLayout
        searchQueryErrorLayout =
                holder.findViewById(R.id.edit_search_string_layout) as TextInputLayout

        engineNameEditText = holder.findViewById(R.id.edit_engine_name) as EditText
        engineNameEditText?.addTextChangedListener(
            buildTextWatcherForErrorLayout(
                engineNameErrorLayout!!
            )
        )
        searchQueryEditText = holder.findViewById(R.id.edit_search_string) as EditText
        searchQueryEditText?.addTextChangedListener(
            buildTextWatcherForErrorLayout(
                searchQueryErrorLayout!!
            )
        )

        progressView = holder.findViewById(R.id.progress) as ProgressBar

        updateState()
        if (lastEditTextTouched == null) {
            ViewUtils.showKeyboard(engineNameEditText)
        }
    }

    private fun saveEditText() {
        if (engineNameEditText?.isFocused == true) {
            lastEditTextTouched = SEARCH_ENGINE_NAME_KEY
        } else if (searchQueryEditText?.isFocused == true) {
            lastEditTextTouched = SEARCH_QUERY_KEY
        }
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        val bundle = state as Bundle
        super.onRestoreInstanceState(bundle.getParcelable(SUPER_STATE_KEY))
        savedSearchEngineName = bundle.getString(SEARCH_ENGINE_NAME_KEY)
        savedSearchQuery = bundle.getString(SEARCH_QUERY_KEY)
        lastEditTextTouched = bundle.getString(LASTEDITTEXT)
        cursorPosition = bundle.getString(CURSOR_POSITION)?.toInt()
    }

    override fun onSaveInstanceState(): Parcelable {
        val state = super.onSaveInstanceState()
        saveEditText()
        val bundle = Bundle()
        bundle.putParcelable(SUPER_STATE_KEY, state)
        bundle.putString(SEARCH_ENGINE_NAME_KEY, engineNameEditText?.text.toString())
        bundle.putString(SEARCH_QUERY_KEY, searchQueryEditText?.text.toString())
        bundle.putString(LASTEDITTEXT, lastEditTextTouched)
        bundle.putString(CURSOR_POSITION, getCursorPosition(lastEditTextTouched))
        return bundle
    }

    private fun getCursorPosition(lastEditTxtSelection: String?): String? = when (lastEditTxtSelection) {
        SEARCH_ENGINE_NAME_KEY -> engineNameEditText?.selectionStart.toString()
        SEARCH_QUERY_KEY -> searchQueryEditText?.selectionStart.toString()
        else -> (-1).toString()
    }

    fun validateEngineNameAndShowError(engineName: String): Boolean {
        val errorMessage = when {
            TextUtils.isEmpty(engineName) ->
                context.getString(R.string.search_add_error_empty_name)

            else -> null
        }

        engineNameErrorLayout?.error = errorMessage
        return errorMessage == null
    }

    fun validateSearchQueryAndShowError(searchQuery: String): Boolean {
        val errorMessage = when {
            TextUtils.isEmpty(searchQuery) -> context.getString(R.string.search_add_error_empty_search)
            !UrlUtils.isValidSearchQueryUrl(searchQuery) -> context.getString(R.string.search_add_error_format)
            else -> null
        }

        searchQueryErrorLayout?.error = errorMessage
        return errorMessage == null
    }

    fun setSearchQueryErrorText(err: String) {
        searchQueryErrorLayout?.error = err
    }

    fun setProgressViewShown(isShown: Boolean) {
        progressView?.visibility = if (isShown) View.VISIBLE else View.GONE
    }

    private fun updateState() {
        if (engineNameEditText != null) engineNameEditText?.setText(savedSearchEngineName)
        if (searchQueryEditText != null) searchQueryEditText?.setText(savedSearchQuery)
        if (lastEditTextTouched != null) {
            if (lastEditTextTouched == SEARCH_ENGINE_NAME_KEY) {
                engineNameEditText?.requestFocus()
                if (engineNameEditText != null && cursorPosition != null) {
                    setCursorPosition(engineNameEditText!!, cursorPosition!!)
                }
            } else if (lastEditTextTouched == SEARCH_QUERY_KEY) {
                searchQueryEditText?.requestFocus()
                if (searchQueryEditText != null && cursorPosition != null) {
                    setCursorPosition(searchQueryEditText!!, cursorPosition!!)
                }
            }
        }
    }

    private fun setCursorPosition(editText: EditText, position: Int) {
        editText.setSelection(position)
    }

    private fun buildTextWatcherForErrorLayout(errorLayout: TextInputLayout): TextWatcher {
        return object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                errorLayout.error = null
            }

            override fun afterTextChanged(editable: Editable) {}
        }
    }

    companion object {
        private val SUPER_STATE_KEY = "super-state"
        private val SEARCH_ENGINE_NAME_KEY = "search-engine-name"
        private val SEARCH_QUERY_KEY = "search-query"
        private const val LASTEDITTEXT = "save-edit-text"
        private const val CURSOR_POSITION = "cursor-position"
    }
}
